File: 4075-count-subarrays-with-majority-element-ii/count-subarrays-with-majority-element-ii.py
class Solution:
    def countMajoritySubarrays(self, nums: List[int], target: int) -> int:
        pre = defaultdict(int)
        pre[0] = 1

        currentPrefixSum = 0
        validEndingHere = 0
        ans = 0

        for x in nums:
            if x == target:
                validEndingHere += pre[currentPrefixSum]
                currentPrefixSum += 1
                pre[currentPrefixSum] += 1
            else:
                currentPrefixSum -= 1
                validEndingHere -= pre[currentPrefixSum]
                pre[currentPrefixSum] += 1

            ans += validEndingHere

        return ans

File: 4075-count-subarrays-with-majority-element-ii/count-subarrays-with-majority-element-ii.cpp
class Solution {
public:
    long long countMajoritySubarrays(vector<int>& nums, int target) {
        unordered_map<int, long long> pre;
        pre[0] = 1;

        int currentPrefixSum = 0;
        long long validEndingHere = 0;
        long long ans = 0;

        for (int num : nums) {
            if (num == target) {
                validEndingHere += pre[currentPrefixSum];

                currentPrefixSum++;
                pre[currentPrefixSum]++;
            } else {
                currentPrefixSum--;

                validEndingHere -= pre[currentPrefixSum];
                pre[currentPrefixSum]++;
            }

            ans += validEndingHere;
        }

        return ans;
    }
};